File: app/src/main/java/com/thetechnocafe/gurleensethi/bottomsheets/BottomSheetActivity.java
package com.thetechnocafe.gurleensethi.bottomsheets;

import android.support.annotation.NonNull;
import android.support.design.widget.BottomSheetBehavior;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

public class BottomSheetActivity extends AppCompatActivity {

    private BottomSheetBehavior mBottomSheetBehaviour;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_bottom_sheet);

        Button openBottomSheetButton = (Button) findViewById(R.id.openBottomSheetButton);
        Button closeBottomSheetButton = (Button) findViewById(R.id.closeBottomSheetButton);

        View nestedScrollView = (View) findViewById(R.id.nestedScrollView);
        mBottomSheetBehaviour = BottomSheetBehavior.from(nestedScrollView);

        //Remove this line to disable peek
        mBottomSheetBehaviour.setPeekHeight(200);

        openBottomSheetButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mBottomSheetBehaviour.setState(BottomSheetBehavior.STATE_EXPANDED);
            }
        });

        closeBottomSheetButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mBottomSheetBehaviour.setState(BottomSheetBehavior.STATE_COLLAPSED);
            }
        });

        mBottomSheetBehaviour.setBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {
                String state = "";

                switch (newState) {
                    case BottomSheetBehavior.STATE_DRAGGING: {
                        state = "DRAGGING";
                        break;
                    }
                    case BottomSheetBehavior.STATE_SETTLING: {
                        state = "SETTLING";
                        break;
                    }
                    case BottomSheetBehavior.STATE_EXPANDED: {
                        state = "EXPANDED";
                        break;
                    }
                    case BottomSheetBehavior.STATE_COLLAPSED: {
                        state = "COLLAPSED";
                        break;
                    }
                    case BottomSheetBehavior.STATE_HIDDEN: {
                        state = "HIDDEN";
                        break;
                    }
                }

                Toast.makeText(BottomSheetActivity.this, "Bottom Sheet State Changed to: " + state, Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {

            }
        });
    }
}

File: app/src/main/java/com/thetechnocafe/gurleensethi/bottomsheets/BottomSheetActivity.kt
package com.thetechnocafe.gurleensethi.bottomsheets

import android.os.Bundle
import android.support.design.widget.BottomSheetBehavior
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_bottom_sheet.*

class BottomSheetActivity : AppCompatActivity() {

    private var mBottomSheetBehaviour: BottomSheetBehavior<*>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_bottom_sheet)

        mBottomSheetBehaviour = BottomSheetBehavior.from(nestedScrollView)

        //Remove this line to disable peek
        mBottomSheetBehaviour!!.peekHeight = 200

        openBottomSheetButton.setOnClickListener { mBottomSheetBehaviour!!.state = BottomSheetBehavior.STATE_EXPANDED }

        closeBottomSheetButton.setOnClickListener { mBottomSheetBehaviour!!.state = BottomSheetBehavior.STATE_COLLAPSED }

        mBottomSheetBehaviour!!.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                var state = ""

                when (newState) {
                    BottomSheetBehavior.STATE_DRAGGING -> {
                        state = "DRAGGING"
                    }
                    BottomSheetBehavior.STATE_SETTLING -> {
                        state = "SETTLING"
                    }
                    BottomSheetBehavior.STATE_EXPANDED -> {
                        state = "EXPANDED"
                    }
                    BottomSheetBehavior.STATE_COLLAPSED -> {
                        state = "COLLAPSED"
                    }
                    BottomSheetBehavior.STATE_HIDDEN -> {
                        state = "HIDDEN"
                    }
                }

                Toast.makeText(this@BottomSheetActivity, "Bottom Sheet State Changed to: $state", Toast.LENGTH_SHORT).show()
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) {

            }
        })
    }
}
